src/bin/pg_autoctl/cli_service.c

```diff
@@ -238,6 +211,50 @@ cli_monitor_run(int argc, char **argv)
 }
 
 
+/*
+ * ensure_monitor_running checks if monitor is running, attempts to restart it if it is not.
+ * verifies the extension version is the same as expected and returns the postgres uri for the
+ * monitor connection upon success.
+ */
+static void
+ensure_monitor_running(Monitor *monitor, MonitorConfig *mconfig)
+{
+	MonitorExtensionVersion version = { 0 };
+	char postgresUri[MAXCONNINFO];
+
+	if (!pg_is_running(mconfig->pgSetup.pg_ctl, mconfig->pgSetup.pgdata))
+	{
+		log_info("Postgres is not running, starting postgres");
+
+		if (!pg_ctl_start(mconfig->pgSetup.pg_ctl,
+						  mconfig->pgSetup.pgdata,
+						  mconfig->pgSetup.pgport,
+						  mconfig->pgSetup.listen_addresses))
+		{
+			log_error("Failed to start PostgreSQL, see above for details");
+			exit(EXIT_CODE_PGCTL);
+		}
+	}
+
+	log_info("PostgreSQL is running in \"%s\" on port %d",
+			 mconfig->pgSetup.pgdata, mconfig->pgSetup.pgport);
+
+	/* Check version compatibility */
+	if (!monitor_ensure_extension_version(monitor, &version))
+	{
+		/* errors have already been logged */
+		exit(EXIT_CODE_MONITOR);
+	}
```

Comment:
I think we should only do this when we restart Postgres, no need to do that check at every loop, right? Also, again, I don't think we should exit in a lower level function.